File: C++/프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.cpp
#include <vector>
#include <queue>

using namespace std;

int solution(int bridge_length, int weight, vector<int> truck_weights) {
    int answer = 0, cur_weight = 0, truck_index = 0, queue_size;
    
    queue<pair<int, int>> q;
    q.push({0, bridge_length});
    pair<int, int> truck;
    
    while(!q.empty()) {
        answer++;
        // cout << "경과시간: " << answer << endl;
        queue_size = q.size();
        for(int i = 0; i < queue_size; i++)  {
            truck = q.front();
            q.pop();
            if(truck.second + 1 >= bridge_length) { // 다리를 다 지났을 때
                cur_weight -= truck.first;
            } else { //다리를 건너고 있을 때
                q.push({truck.first, truck.second + 1});
            }
        }
        
        if(truck_index < truck_weights.size() && cur_weight + truck_weights[truck_index] <= weight && q.size() <= bridge_length) {
            cur_weight += truck_weights[truck_index];
            q.push({truck_weights[truck_index++], 0});
        }
    }
    return answer;
}

File: Python3/프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.py
from collections import deque

def solution(bridge_length, weight, truck_weights):
    i, time, cur_weight = 0, 0, 0
    queue, truck, dist = deque(), 0, 0
    queue.append([0, bridge_length])
    
    while queue:
        time += 1
        # print("time:", time, "/ queue:", queue)
        for q in range(len(queue)):
            truck, dist = queue.popleft()
            if dist + 1 < bridge_length:
                queue.append([truck, dist + 1])
            else:
                cur_weight -= truck
        if len(queue) < bridge_length and i < len(truck_weights) and cur_weight + truck_weights[i] <= weight:
            cur_weight += truck_weights[i]
            queue.append([truck_weights[i], 0])
            i += 1
    
    
    return time